python/ray/data/_internal/plan.py

```diff
@@ -449,16 +424,16 @@ def meta_count(self) -> Optional[int]:
         Returns:
             The number of records of the result Dataset, or None.
         """
-        if self.needs_eager_execution():
-            # If the plan is input/read only, we execute it, so snapshot has output.
-            # This applies to newly created dataset. For example, initial dataset
-            # from read, and output datasets of Dataset.split().
-            self.execute()
-
-        if self._logical_plan.dag is self._snapshot_operator:
-            return self._get_num_rows_from_blocks_metadata(self._snapshot_blocks)
+        if (
+            self._snapshot_bundle is not None
+            and not self._snapshot_operator.output_dependencies
+        ):
+            num_rows = sum(m.num_rows for m in self._snapshot_bundle.metadata)
```

Comment:
nit: should we refactor `_get_num_rows_from_blocks_metadata()` to take RefBundles / block metadata, and combine code usages like `sum(m.num_rows for m in self._snapshot_bundle.metadata)`? otherwise, we could also drop `_get_num_rows_from_blocks_metadata`

Reply:
Yeah, I removed `_get_num_rows_from_blocks_metadata` in https://github.com/ray-project/ray/pull/45860